src/t_list.c

```diff
@@ -394,28 +394,78 @@ void listElementsRemoved(client *c, robj *key, int where, robj *o) {
         dbDelete(c->db, key);
     }
     signalModifiedKey(c, c->db, key);
-    server.dirty++;
+    server.dirty += count;
 }
 
+/* Implements the generic list pop operation for LPOP/RPOP.
+ * The where argument specifies which end of the list is operated on. An
+ * optional count may be provided as the third argument of the client's
+ * command. */
 void popGenericCommand(client *c, int where) {
+    long count = 0;
+    robj *value;
+
+    /* Parse the optional count argument. */
+    if (c->argc == 3) {
+        if (getPositiveLongFromObjectOrReply(c,c->argv[2],&count,NULL) != C_OK) 
+            return;
+        if (count == 0) {
+            /* Fast exit path. */
+            addReplyNullArray(c);
+            return;
+        }
+    }
+
     robj *o = lookupKeyWriteOrReply(c, c->argv[1], shared.null[c->resp]);
     if (o == NULL || checkType(c, o, OBJ_LIST))
         return;
 
-    robj *value = listTypePop(o, where);
-    if (value == NULL) {
-        addReplyNull(c);
-    } else {
+    if (!count) {
+        /* Pop a single element. */
+        value = listTypePop(o,where);
+        serverAssert(value != NULL);
         addReplyBulk(c,value);
         decrRefCount(value);
-        listElementsRemoved(c,c->argv[1],where,o);
+        listElementsRemoved(c,c->argv[1],where,o,1);
+    } else {
+        /* Pop a range of elements. */
+        long llen = listTypeLength(o);
+        long rangelen = (count > llen) ? llen : count;
+        long delcount = rangelen;
+        long rangestart = (where == LIST_HEAD) ? 0 : llen - 1;
+        long delstart = (where == LIST_HEAD) ? 0 : llen - delcount;
+        int direction = (where == LIST_HEAD) ? LIST_TAIL : LIST_HEAD;
+
+        addReplyArrayLen(c,rangelen);
+        if (o->encoding == OBJ_ENCODING_QUICKLIST) {
+            listTypeIterator *iter = listTypeInitIterator(o,rangestart,direction);
+
+            while(rangelen--) {
+                listTypeEntry entry;
+                listTypeNext(iter, &entry);
+                quicklistEntry *qe = &entry.entry;
+                if (qe->value) {
+                    addReplyBulkCBuffer(c,qe->value,qe->sz);
+                } else {
+                    addReplyBulkLongLong(c,qe->longval);
+                }
+            }
+            listTypeReleaseIterator(iter);
```

Comment:
nit: This code overlaps with lrange, would be nice to generalize.

Reply:
Corrected via a13da02 and added a test to preserve the old behavior, which is a shame because `LRANGE l -2 -1` should work imo but will be a total deal breaker.